hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/dsl/operations/transactions/ApproveAllowanceOperation.java

```diff
@@ -0,0 +1,117 @@
+
+import static com.hedera.services.bdd.spec.transactions.TxnVerbs.cryptoApproveAllowance;
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.services.bdd.spec.HapiSpec;
+import com.hedera.services.bdd.spec.SpecOperation;
+import com.hedera.services.bdd.spec.dsl.entities.SpecAccount;
+import com.hedera.services.bdd.spec.dsl.entities.SpecContract;
+import com.hedera.services.bdd.spec.dsl.entities.SpecToken;
+import com.hedera.services.bdd.spec.transactions.crypto.HapiCryptoApproveAllowance;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Adds option for approving allowance for fungible, non-fungible tokens and hBars as well.
+ */
+public class ApproveAllowanceOperation
+        extends AbstractSpecTransaction<ApproveAllowanceOperation, HapiCryptoApproveAllowance>
+        implements SpecOperation {
+
+    @NonNull
+    private final String ownerName;
+
+    @NonNull
+    private final String spenderName;
+
+    @Nullable
+    private final SpecToken token;
+
+    private final long allowance;
+    private final List<Long> serialNumbers = new ArrayList<>();
+    private boolean isApproveForAll;
+
+    public ApproveAllowanceOperation(
+            @NonNull final SpecToken token,
+            @NonNull final SpecAccount owner,
+            @NonNull final SpecContract spender,
+            final long amount) {
+        super(List.of(token, owner, spender));
+
+        this.allowance = amount;
+        this.ownerName = requireNonNull(owner.name());
+        this.spenderName = spender.name();
+        this.token = requireNonNull(token);
+    }
+
+    public ApproveAllowanceOperation(
+            @NonNull final SpecAccount owner, @NonNull final SpecContract spender, final long amount) {
+        super(List.of(owner, spender));
+
+        this.allowance = amount;
+        this.ownerName = requireNonNull(owner.name());
+        this.spenderName = requireNonNull(spender.name());
+        this.token = null;
+    }
+
+    public ApproveAllowanceOperation(
+            @NonNull final SpecToken token,
+            @NonNull final SpecAccount owner,
+            @NonNull final SpecContract spender,
+            final boolean isApproveForAll,
+            @NonNull final List<Long> serialNumbers) {
+        super(List.of(token, owner, spender));
+
+        this.allowance = 0;
+        this.ownerName = requireNonNull(owner.name());
+        this.spenderName = requireNonNull(spender.name());
+        this.token = requireNonNull(token);
+        this.serialNumbers.addAll(serialNumbers);
+        this.isApproveForAll = isApproveForAll;
+    }
+
+    @Override
+    protected ApproveAllowanceOperation self() {
+        return this;
+    }
+
+    @NonNull
+    @Override
+    protected SpecOperation computeDelegate(@NonNull final HapiSpec spec) {
+        /* Approve allowance for hBar */
```

Comment:
I wonder if it would be clearer if there were an enum field that would list the various approval types.  This would be a simple switch statement in that case without having to read code.

Reply:
I tested it out, it definitely looks better with switch case, so I pushed the changes. :)  
Thanks for the suggestion.